streams/shared/src/main/scala/zio/stream/StreamEffect.scala

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.stream
+
+import zio._
+
+private[stream] trait StreamEffect[+E, +A] extends ZStream[Any, E, A] { self =>
+
+  def processEffect: Managed[E, () => A]
+
+  override def collect[B](pf: PartialFunction[A, B]): StreamEffect[E, B] =
+    new StreamEffect[E, B] {
+      def process = StreamEffect.super.collect(pf).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            def pull(): B = pf.applyOrElse(it(), (_: A) => pull())
+            () => pull()
+          }
+        }
+    }
+
+  override def collectWhile[B](pred: PartialFunction[A, B]): StreamEffect[E, B] =
+    new StreamEffect[E, B] {
+      def process = StreamEffect.super.collectWhile(pred).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            var done = false
+
+            () => {
+              if (done) throw StreamEffect.End
+              else pred.applyOrElse(it(), (_: A) => { done = true; throw StreamEffect.End })
+            }
+          }
+        }
+    }
+
+  override def dropWhile(pred: A => Boolean): StreamEffect[E, A] =
+    new StreamEffect[E, A] {
+      def process = StreamEffect.super.dropWhile(pred).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            var drop = true
+
+            def pull(): A = {
+              val a = it()
+              if (!drop) a
+              else if (!pred(a)) {
+                drop = false
+                a
+              } else pull()
+            }
+
+            () => pull()
+          }
+        }
+    }
+
+  override def filter(pred: A => Boolean): StreamEffect[E, A] =
+    new StreamEffect[E, A] {
+      def process = StreamEffect.super.filter(pred).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            def pull(): A = {
+              val a = it()
+              if (pred(a)) a else pull()
```

Comment:
This needs a `tailrec` annotation, which I am not sure will work reliably when nested like this (the `scalac` bug may be fixed though).

Reply:
This is done already and compiling without issues on all Scala versions.